Common input file processing for CLI and Standard JSON

cameel opened the pull request:
Fixes #11552.
~Depends on #11520. Draft until that one is merged.~ Merged.

The main purpose of this refactor is to group input file processing and `FileReader` initialization in `CommandLineInterface` into an independent phase (`readInputFiles()`), separate from actually running the compiler. This way its results can be easily tested. So far almost all of this was already grouped into `readInputFilesAndConfigureFileReader()` except for stuff related to Standard JSON. This PR unifies the logic for Standard JSON with that for other modes.

As a side effect this improves the behavior with `--standard-input`:
- Fixes the crash from #11552.
-  An explicit request to read from standard input (the `-` option) now works in that mode too.
-  Technically, `--ignore-missing` now works as well but it's useless because in Standard JSON you cannot specify more than one file.
- If you try to specify remappings on the command line you get an error explicitly telling you to put them in JSON instead of one saying that there are too many input fiies.

cameel committed:
CommandLineParser: Refactor checkMutuallyExclusive() to handle multiple options and use it more

cameel committed:
CommandLineParser: Initialize inputMode earlier to make it usable in more functions

cameel committed:
Process input files in Standard JSON mode just like in other modes

- This makes `-` for stdin work.
- `--ignore-missing` now works with `--standard-json`, though it's not very useful because there can be at most one input file.
- Separate errors for situations where there are no input files on the command line (this can be detected in the parser) and where they are not present on disk.

christianparpart commented on solc/CommandLineInterface.cpp:
```diff
@@ -377,8 +377,18 @@ void CommandLineInterface::handleGasEstimation(string const& _contract)
-bool CommandLineInterface::readInputFilesAndConfigureFileReader()
+bool CommandLineInterface::readInputFiles()
 {
+	solAssert(!m_standardJsonInput.has_value(), "");
```
```suggestion
	solAssert(!m_standardJsonInput, "");
```

cameel replied:
I find `!x.has_value()` much more readable than `!x`. The latter often does not really read like a condition. And it makes `x` look like a pointer. I'd prefer to keep this as is.

christianparpart replied:
I am not sure if we should come up with something as a rule-of-thumb to this question, because I see it written in both ways, and I am getting corrected into at least one way (using `operator!` and `operator bool()`). I'm fine with either way.

cameel replied:
Yeah, one of these things where we'd need to agree on common style. Until then you probably will get corrected either way :)

BTW, I saw @ekpyron using `has_value()` in his PR so it's not just me :)

ekpyron replied:
Although I mainly use it, because people are telling me it's nicer and I don't really mind either way myself :-). Unfortunately though, as a consequence, I don't think I do it consistently either way myself these days :-).

cameel replied:
> because people are telling me it's nicer

Well, they're right :P

I'm not really sure why people are not bothered more by using non-boolean values in conditions but for me it just makes them harder to read so it's not worth saving a few keystrokes.

ekpyron replied:
I mainly use non-boolean values on conditions in branches, like ``if (auto someOptional = someFunction())``.
You could write ``if (auto someOptional = someFunction(); someOptional.has_value())`` for those, but in that case it seems like a pretty weird thing to do :-). In all other cases I could get used to using ``.has_value()``. But what do you propose for checking a pointer being not-``nullptr``? That's also a non-boolean value... we do have cases of ``!!ptr`` for that, but IMHO that doesn't really improve anything :-). And ``ptr != nullptr`` seems overly verbose.

ekpyron replied:
And if it's fine for a pointer to just implicitly have it convert to `bool`, I'm not sure why it should be wrong for an `optional`...

cameel replied:
I don't see anything wrong with `ptr != nullptr`. It reads like a full condition (if ptr is not null). `ptr` as a condition is just weird and I only use it because it's prevalent style in our code already. Otherwise I'd never do that.

I don't think it's wrong. I just think it's not as readable. This style is really common in Python and I really hate it.

> You could write `if (auto someOptional = someFunction(); someOptional.has_value())` for those, but in that case it seems like a pretty weird thing to do :-).

Agreed. I mean, I'd prefer to have has_value()` there but a preference for not having the second part in the `if` is much stronger :) It's too easy to overlook that part when skimming over code.

ekpyron replied:
Actually, e.g. in ``assert(ptr != nullptr)`` or even ``if(ptr != nullptr)`` I'd even be ok with it - I'm just extremely used to implicit pointer-to-bool conversions, so I'd probably have to be constantly reminded until I'm used to it :-D. But even more prevalent in the code is again the ``if (auto ptr = ...sth...)`` case, especially with the RTTI ``dynamic_cast`` cases... and there ``if (auto ptr = ...; ptr != nullptr)`` has the same problem as the optional case... but yeah, not sure. Maybe we should more broadly collect opinions. *Sometimes* using ``assert(!!ptr)``, *sometimes* ``assert(ptr)``, *sometimes* ``assert(someOptional)`` and *sometimes* ``assert(someOptional.has_value())`` randomly for sure isn't the nicest thing to do.

christianparpart commented on solc/CommandLineInterface.cpp:
```diff
@@ -411,16 +421,33 @@ bool CommandLineInterface::readInputFilesAndConfigureFileReader()
+		string fileContent = readFileAsString(infile.string());
+		if (m_options.input.mode == InputMode::StandardJson)
+		{
+			solAssert(!m_standardJsonInput.has_value(), "");
```
```suggestion
			solAssert(!m_standardJsonInput, "");
```

christianparpart commented on solc/CommandLineInterface.cpp:
```diff
@@ -411,16 +421,33 @@ bool CommandLineInterface::readInputFilesAndConfigureFileReader()
 		// NOTE: we ignore the FileNotFound exception as we manually check above
-		m_fileReader.setSource(infile, readFileAsString(infile.string()));
-		m_fileReader.allowDirectory(boost::filesystem::path(boost::filesystem::canonical(infile).string()).remove_filename());
+		string fileContent = readFileAsString(infile.string());
```
it's called "read**File**AsString", maybe it should expect a filesystem::path then? So you don't have to pass `infile.string()` but only `infile`. (just a code cleanup idea).

cameel replied:
Right, this should be a path. It's unrelated to this PR though so I have submitted a separate one: #11650.

The downside is that now `CommonIO.h` includes the boost path header but I guess we can live with that.

christianparpart commented on solc/CommandLineInterface.cpp:
```diff
@@ -411,16 +421,33 @@ bool CommandLineInterface::readInputFilesAndConfigureFileReader()
+		else
+		{
+			m_fileReader.setSource(infile, move(fileContent));
+			m_fileReader.allowDirectory(boost::filesystem::path(boost::filesystem::canonical(infile).string()).remove_filename());
```
```suggestion
			m_fileReader.allowDirectory(boost::filesystem::path(boost::filesystem::canonical(infile)).remove_filename());
```
This should work, too, isn't it?

cameel replied:
Yeah. Also the explicit `boost::filesystem::path` conversion is unnecessary.

christianparpart commented on solc/CommandLineInterface.cpp:
```diff
@@ -411,16 +421,33 @@ bool CommandLineInterface::readInputFilesAndConfigureFileReader()
+	{
+		if (m_options.input.mode == InputMode::StandardJson)
+		{
+			solAssert(!m_standardJsonInput.has_value(), "");
```
```suggestion
			solAssert(!m_standardJsonInput, "");
```

christianparpart commented on solc/CommandLineInterface.cpp:
```diff
@@ -411,16 +421,33 @@ bool CommandLineInterface::readInputFilesAndConfigureFileReader()
+	}
 
-	if (m_fileReader.sourceCodes().size() == 0)
+	if (m_fileReader.sourceCodes().size() == 0 && !m_standardJsonInput.has_value())
```
```suggestion
	if (m_fileReader.sourceCodes().empty() && !m_standardJsonInput)
```

cameel replied:
Changed `size() == 0` to `empty()`.

christianparpart commented on solc/CommandLineParser.cpp:
```diff
@@ -899,10 +915,8 @@ General Information)").c_str(),
 	}
 
-	if (m_args.count(g_strAssemble) || m_args.count(g_strStrictAssembly) || m_args.count(g_strYul))
+	if (m_options.input.mode == InputMode::Assembler)
```
You chain the `if`-statements to iterate through each `InputMode` enum class' value.
If you use a switch statement here, the compiler would tell you if you accidentally missed handling one case. I also think it would be cleaner, but that might be subjective. :)

cameel replied:
Right. I guess it's time to start applying the refactors requested in the CLI PR :)
Done.

christianparpart commented on test/solc/CommandLineInterface.cpp:
```diff
@@ -0,0 +1,294 @@
+	size_t argc = commandLine.size();
+	vector<char const*> argv(commandLine.size() + 1);
+
+	// argv[argc] typically contains NULL
```
not just typically. isn't it always?

cameel replied:
Oh. Looks like it's the language standard that mandates this. I though it was just a convention.
Comment fixed.

christianparpart commented on test/solc/CommandLineInterface.cpp:
```diff
@@ -0,0 +1,294 @@
+	// argv[argc] typically contains NULL
+	argv[argc] = nullptr;
+
+	for (size_t i = 0; i < argc; ++i)
```
```suggestion
	for (size_t const i: views::iota(argc))
```
Just a suggestion (I try to avoid mutating variables that don't have any special purpose) :)
This comes from `<ranges/views/iota.hpp>`.

cameel replied:
I get the idea but in this case I'd rather keep the loop as is. The current form is the bog standard idiomatic one that everyone recognizes at a glance. The one with `iota` is still simple but it would still give me a pause because it's not very common :)

christianparpart commented on test/solc/CommandLineInterface.cpp:
```diff
@@ -0,0 +1,294 @@
+	BOOST_TEST(result.stderrContent == expectedMessage);
+}
+
+BOOST_AUTO_TEST_SUITE_END()
```
I really like the testability of the CLI arguments! :+1:

christianparpart commented on test/solc/CommandLineParser.cpp:
```diff
@@ -388,10 +388,10 @@ BOOST_AUTO_TEST_CASE(standard_json_mode_options)
 	expectedOptions.input.basePath = "/home/user/";
 	expectedOptions.input.allowedDirectories = {"/tmp", "/home", "project", "../contracts"};
+	expectedOptions.input.ignoreMissingFiles = true;
 	expectedOptions.output.dir = "/tmp/out";
```
It was already there, but it should never be hardcoded to `/tmp`. Is this used and written to somewhere? (if not, ignore this message). :)

cameel replied:
`CommandLineParser` is not meant to read anything from disk so it's fine. Note that all the actual file reading and checks for path existence are in `CommandLineInterface`. This is for example the reason why both have separate conditions to check if there were any input files - even if `CommandLineParser` checks it and sees that there's at least one path, `CommandLineInterface` with `--ignore-missing` could still see zero files because they do not actually exist.

I decided to use these paths like this because dealing with absolute paths when you need to ensure they exist overcomplicates tests a lot. See for example tests like `cli_paths_to_source_unit_names_relative_base_path` from `test/solc/CommandLineInterface.cpp` in #11545. These tests would be much simpler if I could just write paths like this but there I can't.

cameel commented:
This is now ready for another round.

christianparpart commented on solc/CommandLineInterface.cpp:
```diff
@@ -377,8 +377,18 @@ void CommandLineInterface::handleGasEstimation(string const& _contract)
+
+	if (m_options.input.basePath != "" && !boost::filesystem::is_directory(m_options.input.basePath))
+	{
+		serr() << "Base path must be a directory: \"" << m_options.input.basePath << "\"\n";
```
maybe quote `basePath` here in the output to avoid accidentally misleading output? - Or don't quote at all.

cameel replied:
OK. I tried to be consistent about quoting but looks like existing messages are themselves not consistent :) True, the ones in this part do not quote stuff.

We should really do #7578 some day :)

cameel replied:
Quotes removed.

I actually also changed `m_options.input.basePath` to `m_fileReader.basePath()` because that's what we should be checking.

christianparpart commented on solc/CommandLineInterface.cpp:
```diff
@@ -411,16 +421,33 @@ bool CommandLineInterface::readInputFilesAndConfigureFileReader()
+	if (m_fileReader.sourceCodes().empty() && !m_standardJsonInput.has_value())
 	{
-		serr() << "No input files given. If you wish to use the standard input please specify \"-\" explicitly." << endl;
+		serr() << "All specified input files either do not exist or are not regular files." << endl;
```
Why actually `endl` and not `\n`? Not that we care about performance at that stage (`endl` is flushing, `\n` is not necessarily), but `\n` works on all platforms, and is shorter to write than always this concatenation style for creating a newline.

cameel replied:
Mostly because it's the prevalent style used in this file and I don't think the difference matters for us here. As for readability, I slightly prefer `endl` because `\n` looks like noise to me :) but I don't feel too strongly about either.

If we want to change this, we should go over the whole file (and maybe more files) so I'd rather not do it in this PR.
It might actually be a `good first issue` for someone so maybe create an issue for it?

cameel replied:
Oh, I just noticed that I used `\n` in the base path message (and it's the only one like that in the file). Changed that to `endl` now for consistency.

christianparpart reviewed (dismissed):
LGTM (apart from the minors which seem subjective anyways :)

aarlt commented on solc/CommandLineParser.cpp:
```diff
@@ -846,42 +872,32 @@ General Information)").c_str(),
 	}
 
-	vector<string> const exclusiveModes = {
+	checkMutuallyExclusive({
```
I guess the return value was supposed to be used here.

cameel replied:
Right. Great catch! I'm actually surprised it's not covered by tests. I'll add one.

cameel replied:
Done.

aarlt commented on test/solc/CommandLineParser.cpp:
```diff
@@ -388,10 +388,10 @@ BOOST_AUTO_TEST_CASE(standard_json_mode_options)
 	expectedOptions.input.basePath = "/home/user/";
 	expectedOptions.input.allowedDirectories = {"/tmp", "/home", "project", "../contracts"};
+	expectedOptions.input.ignoreMissingFiles = true;
 	expectedOptions.output.dir = "/tmp/out";
```
I'm not entirely sure, but wouldn't it be good to also add some tests here that test command line parameters that are usually used in different platforms, e.g. windows. Just to verify that the `CommandLineParser` parses e.g. windows specific paths correctly.

cameel replied:
True and I actually do have such tests but only for `CommandLineInterface` (in #11545). `CommandLineInterface` actually reads files and uses these paths. `CommandLineParser` only puts paths in the options struct so I think there's not much gain from repeating these tests for it. Especially given that these tests are pretty bulky and often need ifdefs for other platforms.

Actually, I could move the first commit from #11545 that contains them here but I'm not sure if this makes sense now that it's been reviewed. Originally I wanted it to be a separate PR but I already have too many PRs that depend on each other and I decided to avoid creating more.

aarlt commented:
Looks good to me, at least for now I only found one small thing. 

However, I was wondering whether we should reduce the line numbers used in `bool CommandLineParser::parse(...)`. I think this function is a bit too long (ca. 700 lines).

cameel commented:
> However, I was wondering whether we should reduce the line numbers used in `bool CommandLineParser::parse(...)`. I think this function is a bit too long (ca. 700 lines).

I totally agree. This is not a new thing though. The functions in `CommandLineInterface` were already pretty big and I just did not split them to keep the diff smaller. I'm going to do some more refactors, but step by step, over time because this is not a priority right now. The main point of my refactor was only to make the CLI somewhat testable, not to rewrite it all :)

cameel commented:
Fixes applies, comments answered. This is ready for another round.

cameel commented:
I'm adding some refactors from work on #11688. All of this basically only touches the test utilities and should not require re-reviewing the whole PR. Since this PR has not been merged yet, I think it makes more sense to change this stuff here instead of having to immediately refactor it in later PRs. Everything that has been added is in fixup commits:
- Added `test/solc/Common.h` and moved helpers from `CommandLineInterface`/`CommandLineParser` tests there.
    - Made `parseCommandLineAndReadInputFiles()` a bit more generic.
- Operators for printing sources, allowed paths and remappings if a test fails.
- `createEmptyFilesWithParentDirs()` is now `createFilesWithParentDirs()` and you can specify file content.
- `TEST_CASE_NAME` (see #11686).

christianparpart reviewed (dismissed):
LGTM (again, actually) :)

cameel commented:
Great. In that case I've squashed the fixups and rebased on develop so that it's ready to merge.

cameel commented:
Conflict resolved. It was just a change from #11639 that added the formatting parameter to `StandardCompiler`. I just adjusted the constructor invocation in `CommandLineInterface.cpp`.

christianparpart reviewed (approved):
re-approved after fixes.
